contracts/staking/stakeManager/StakeManager.sol

```diff
@@ -491,123 +532,135 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
-        // update stateMerkleTree root for accounts balance on heimdall chain
-        accountStateRoot = stateRoot;
-        _finalizeCommit();
-        return checkSignature(stakePower, reward, voteHash, sigs);
-    }
+        UnstakedValidatorsContext memory unstakeCtx;
+        unstakeCtx.deactivatedValidators = new uint256[](totalStakers);
+
+        for (uint256 i = 0; i < sigs.length; ++i) {
+            address signer = ECVerify.ecrecovery(voteHash, sigs[i]);
 
-    function checkSignature(
-        uint256 checkpointStakePower,
-        uint256 reward,
-        bytes32 voteHash,
-        bytes memory sigs
-    ) internal returns (uint256) {
-        // total voting power
-        uint256 totalStakePower;
-        address lastAdd; // cannot have address(0x0) as an owner
-        for (uint64 i = 0; i < sigs.length; i += 65) {
-            address signer = voteHash.ecrecovery(BytesLib.slice(sigs, i, 65));
-            uint256 validatorId = signerToValidator[signer];
-            // check if signer is staker and not proposer
             if (signer == lastAdd) {
+                // if signer signs twice, just skip this signature
+                continue;
+            }
+
+            if (signer < lastAdd) {
+                // if signatures are out of order - break out, it is not possible to keep track of unsigned validators
                 break;
-            } else if (isValidator(validatorId) && signer > lastAdd) {
+            }
+
+            uint256 validatorId = signerToValidator[signer];
+
+            unsignedCtx = _fillUnsignedValidators(unsignedCtx, signer, unsignedCtx.totalValidators);
+
+            uint256 amount = validators[validatorId].amount;
+            Status status = validators[validatorId].status;
+            
+            unstakeCtx.deactivationEpoch = validators[validatorId].deactivationEpoch;
+
+            if (_isValidator(status, amount, unstakeCtx.deactivationEpoch, _currentEpoch)) {
                 lastAdd = signer;
 
-                Validator storage validator = validators[validatorId];
-                uint256 valPow;
-                // add delegation power
-                address contractAddr = validator.contractAddress;
-                if (contractAddr != address(0x0)) {
-                    valPow = IValidatorShare(contractAddr).updateRewards(
-                        reward,
-                        checkpointStakePower,
-                        validatorStake(validatorId)
-                    );
-                } else {
-                    valPow = validator.amount;
-                    validator.reward = validator.reward.add(valPow.mul(reward).div(checkpointStakePower));
+                signedStakePower = signedStakePower.add(validators[validatorId].delegatedAmount.add(amount));
+
+                if (unstakeCtx.deactivationEpoch != 0) {
+                    unstakeCtx.deactivatedValidators[unstakeCtx.validatorIndex] = validatorId;
```

Comment:
Do we need to handle this explicitly? Because if it is a "leaving" validator and still active, shouldn't it be the same as an active validator?

Reply:
Not really, we must explicitly update his rewards, otherwise, he will not get rewards. UnstakeClaim doesn't update rewards to avoid unstaked validator from getting more rewards.